### ISSUE TITLE
Internal error: list of lists passed to proc

### ISSUE THREAD

**Issue by stonea** [chapel-lang/chapel#17319]
### Summary of Problem
I'm getting the following "internal error" message when compiling a program that calls into a procedure taking a list of lists of ints as an argument.

```
bug.chpl:3: internal error: UTI-MIS-0889 chpl version 1.24.0 pre-release (2b5162ef27)
Note: This source location is a guess.

Internal errors indicate a bug in the Chapel compiler ("It's us, not you"),
and we're sorry for the hassle.  We would appreciate your reporting this bug --
please see https://chapel-lang.org/bugs.html for instructions.  In the meantime,
the filename + line number above may be useful in working around the issue.
```

### Steps to Reproduce

**Source Code:**

```chapel
use List;

proc foo(nums : list(list(int))) : int {
    return 0;
}

proc main() {
    const input = new list(new list(int));
    const actual = foo(input);
}
```

**Compile command:**
`chpl bug.chpl`

### Configuration Information

I cloned the Chapel repos a couple weeks ago and compiled it locally on my machine. I'm running it on Ubuntu on top of Windows Subsystem for Linux.  Output from specific commands of interest are:

```
> chpl --version
chpl version 1.24.0 pre-release (2b5162ef27)
Copyright 2020-2021 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)

> $CHPL_HOME/util/printchplenv --anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none

> gcc --version
gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by e-kayrakli** [comment]
While any error that shows as you report is a bug that we need to fix in the compiler, I think the code in the sample is erroneous and the compiler should report an understandable error.

This line

```chapel
const input = new list(new list(int));
```

should be

```chapel
const input = new list(list(int));
```

where the argument you pass to `new list` must be another type, whereas `new list(int)` as you pass is a value.

**Comment by bradcray** [comment]
Thanks for the quick analysis, Engin.  Do you have a grasp on what a more generalized title for this issue should be?  (e.g., "passing value where type expected causes internal error" seems in the right ballpark, but maybe still too broad?  Unless it's truly that bad.

**Comment by stonea** [comment]
I can reproduce the issue with this simplified example:

```chpl
record GenericRecord {
  type t;
  param p = false;

  proc init(type t, param p=false) {
    this.t = t;
  }
}

proc foo(nums : GenericRecord(GenericRecord(int))) {}
proc main() {
  const actual = new GenericRecord(int);
  foo(actual);
}
```

**Comment by stonea** [comment]
Getting into the weeds here:

Coming out of normalization, `fixupQueryFormals()` and adds a "where" clause (I assume to check if the incoming type matches the expected formal type).  For the example I gave in the previous comment, this is the AST it produces for the where clause:

<details>
<pre>
  (466460)  BlockStmt where
  (466475)    CallExpr
  (466476)      UnresolvedSymExpr '&')
  (466461)      CallExpr
  (466462)        UnresolvedSymExpr '&')
  (466459)        SymExpr 'const-val true[35]:bool[10]')
  (466458)        CallExpr is_instantiation_allow_values
  (466454)          SymExpr 'type GenericRecord[198015]:GenericRecord[198015]?')
  (466455)          CallExpr query
  (466456)            SymExpr 'unknown arg nums[198031]:_unknown[54]')
  (466457)            SymExpr 1 'val _literal_48[6177]:int(64)[13]'))))
  (466473)      CallExpr
  (466474)        UnresolvedSymExpr '==')
  (466472)        SymExpr 'type int(64)[13]:int(64)[13]')
  (466467)        CallExpr query
  (466468)          CallExpr query
  (466469)            SymExpr 'unknown arg nums[198031]:_unknown[54]')
  (466470)            SymExpr 1 'val _literal_48[6177]:int(64)[13]'))
  (466471)          SymExpr 1 'val _literal_48[6177]:int(64)[13]'))))))
</pre>
</details>

Reformatted for readability:

```
  unresolvedSym<&>
      unresolvedSym<&>
          %true,
          is_instantiation_allow_values(%GenericRecord, query(%nums, %1)),
      unresolvedSym<==>(
          %type int(64),
          query(query(%nums, %1), %1)
```

By the time we get to resolution, the AST for the where clause in what I assume is the template function of `foo` looks like this:

<details>
<pre>
  (466460)  BlockStmt where
  (714405)    DefExpr   "expr temp" "maybe param" "maybe type" "temp"'unknown call_tmp[714404]:_unknown[54]')
  (714407)    CallExpr move
  (714408)      SymExpr 'unknown call_tmp[714404]:_unknown[54]')
  (466455)      CallExpr query
  (466456)        SymExpr 'unknown arg nums[198031]:GenericRecord[198015]')
  (466457)        SymExpr 1 'val _literal_48[6177]:int(64)[13]')))
  (714410)    DefExpr   "expr temp" "maybe param" "maybe type" "temp"'unknown call_tmp[714409]:_unknown[54]')
  (714412)    CallExpr move
  (714413)      SymExpr 'unknown call_tmp[714409]:_unknown[54]')
  (466458)      CallExpr is_instantiation_allow_values
  (466454)        SymExpr 'type GenericRecord[198015]:GenericRecord[198015]?')
  (714406)        SymExpr 'unknown call_tmp[714404]:_unknown[54]')))
  (714415)    DefExpr   "expr temp" "maybe param" "maybe type" "temp"'unknown call_tmp[714414]:_unknown[54]')
  (714417)    CallExpr move
  (714418)      SymExpr 'unknown call_tmp[714414]:_unknown[54]')
  (466461)      CallExpr
  (466462)        UnresolvedSymExpr '&')
  (466459)        SymExpr 'const-val true[35]:bool[10]')
  (714411)        SymExpr 'unknown call_tmp[714409]:_unknown[54]')))
  (714420)    DefExpr   "expr temp" "maybe param" "maybe type" "temp"'unknown call_tmp[714419]:_unknown[54]')
  (714422)    CallExpr move
  (714423)      SymExpr 'unknown call_tmp[714419]:_unknown[54]')
  (466468)      CallExpr query
  (466469)        SymExpr 'unknown arg nums[198031]:GenericRecord[198015]')
  (466470)        SymExpr 1 'val _literal_48[6177]:int(64)[13]')))
  (714425)    DefExpr   "expr temp" "maybe param" "maybe type" "temp"'unknown call_tmp[714424]:_unknown[54]')
  (714427)    CallExpr move
  (714428)      SymExpr 'unknown call_tmp[714424]:_unknown[54]')
  (466467)      CallExpr query
  (714421)        SymExpr 'unknown call_tmp[714419]:_unknown[54]')
  (466471)        SymExpr 1 'val _literal_48[6177]:int(64)[13]')))
  (714430)    DefExpr   "expr temp" "maybe param" "maybe type" "temp"'unknown call_tmp[714429]:_unknown[54]')
  (714432)    CallExpr move
  (714433)      SymExpr 'unknown call_tmp[714429]:_unknown[54]')
  (466473)      CallExpr
  (466474)        UnresolvedSymExpr '==')
  (466472)        SymExpr 'type int(64)[13]:int(64)[13]')
  (714426)        SymExpr 'unknown call_tmp[714424]:_unknown[54]')))
  (466475)    CallExpr
  (466476)      UnresolvedSymExpr '&')
  (714416)      SymExpr 'unknown call_tmp[714414]:_unknown[54]')
  (714431)      SymExpr 'unknown call_tmp[714429]:_unknown[54]'))))
</pre>
</details>

Again, reformatted for readability:

```
    %call_tmp_1 <- query(%nums, %1)
    %call_tmp_2 <- is_instantiation_allow_values(%GenericRecord, %call_tmp_1)
    %call_tmp_3 <- &(%true, %call_tmp_2)
    %call_tmp_4 <- query(%nums, %1)
    %call_tmp_5 <- query(%call_tmp_4, %1)
    %call_tmp_6 <- ==(%type int(64), %call_tmp_5)
    &(%call_tmp_3, %call_tmp_6)
```

I guess we do some kind of partial evaluation (constant folding?) at resolution time to evaluate this?  Looking at this AST the immediate thing that jumps out to me is that we don't do any sort of "short circuiting" if the left hand side of the last & (ID: 466475) evaluated to false (where we assign call_tmp_2 in my reformatted AST snippet)

In other words, if my understanding is correct, the check in the where clause is supposed to do the following:

- if the type of the first field of the first formal parameter is a "GenericRecord" then check:
- if the type of that field's first field is int(64)

But we don't stop after the first step evaluates to false (in this example it evaluates to a primitive type) and thus we end up trying to query the first field of a parameter type and naturally thats a non-sensical thing to do and causes an error.

What's really weird to me is that this example seems to depend on having that 'param p'.  I guess a next step is to figure out what AST we're getting out of fixupQueryFormals() there.

**Comment by bradcray** [comment]
@stonea: I was too underwater with release notes during issues week to help with this, though I kept wondering (and still am) whether you ever received any guidance, advice, or information here, or simply timed out and had to move back to "real work"?  I'm also curious whether you feel like you were close enough that we should find someone to help with this to land it (or at least answer unanswered questions for you)?

**Comment by stonea** [comment]
Hi @bradcray,

> whether you ever received any guidance, advice, or information here, or simply timed out and had to move back to "real work"? 

Yeah, unfortunately that's what happened. I probably should have asked for help in the Slack channel at the time.

What I recall with this is that I found what looked like bad IR to me and it runs into trouble later when we try and constant fold it:

https://github.com/chapel-lang/chapel/issues/17319#issuecomment-942834691

> I'm also curious whether you feel like you were close enough that we should find someone to help with this to land it (or at least answer unanswered questions for you)?

I'm thinking the fix could be adding a little extra IR, but I'm not totally sure. It's one of those things where I can point it and say "this looks wrong" and "let's just change the IR to not do that", but there might be some details I'm missing, which make it a lot more complex than what I have in my head.

Anyway, maybe the way to approach this is I could take a 1-pointer on it for the sprint after Thanksgiving. Give myself a chance pull things back into my mental cache and get to the point where I can reproduce the issue in a debugger. Then I could show you (or someone else you can point me to) who's more of an expert in this code and they could advise on whether my diagnoses is correct and where we should go from there.

--

If you have a little bit of time for details --- briefly, what I think happened is I ended up with a piece of IR that looks like this (I've rewritten in my own psuedo-pretty style, hopefully that makes it easier to read not harder):

```
    %call_tmp_1 <- query(%nums, %1)
    %call_tmp_2 <- is_instantiation_allow_values(%GenericRecord, %call_tmp_1)
    %call_tmp_3 <- &(%true, %call_tmp_2)
    %call_tmp_4 <- query(%nums, %1)
    %call_tmp_5 <- query(%call_tmp_4, %1)
    %call_tmp_6 <- ==(%type int(64), %call_tmp_5)
    &(%call_tmp_3, %call_tmp_6)
```

Notice at the end we we do an & of `call_tmp3` and `call_tmp6`. For this specific instantiation when we constant fold it `call_tmp_3` evaluates to false, but we run into some trouble when we do the code that computes `call_tmp6` (or one of its dependencies: `call_tmp4`, `call_tmp5`, or `call_tmp_6`).

I'm thinking the fix may be as simple as just adding a patch of IR to short circuit the computation for `call_tmp_6` when `call_tmp_3` evaluates to false.

I go into more detail in this comment here:

https://github.com/chapel-lang/chapel/issues/17319#issuecomment-942834691
